File: implementations/java-springboot/examples/src/main/java/com/example/accountservice/account/domain/SpendingForecastRepository.java
package com.example.accountservice.account.domain;

public interface SpendingForecastRepository {

    void saveForecast(SpendingForecast forecast);

    /**
     * A cheap idempotency check ahead of the real work — the (accountId, forecastMonth) unique
     * constraint on the table is the last line of defense, the same two-layer pattern as {@link
     * SpendingAnalysisRepository#hasAnalysis}.
     */
    boolean hasForecast(String accountId, String forecastMonth);
}

File: implementations/kotlin-springboot/examples/src/main/kotlin/com/example/accountservice/account/domain/SpendingForecastRepository.kt
package com.example.accountservice.account.domain

/**
 * The write-side port for [SpendingForecast] — a plain data holder rather than a full Aggregate, but
 * still gets its own Repository interface (domain/) + implementation (infrastructure/persistence/)
 * split, the same as [SpendingAnalysisRepository].
 */
interface SpendingForecastRepository {
    fun saveForecast(forecast: SpendingForecast)

    /**
     * A cheap idempotency check ahead of the real work — the (accountId, forecastMonth) unique
     * constraint on the table is the last line of defense, the same two-layer pattern as
     * [SpendingAnalysisRepository.hasAnalysis].
     */
    fun hasForecast(
        accountId: String,
        forecastMonth: String,
    ): Boolean
}
